File: csharp-dotnet/src/Architecture.Shell.Cqrs/IMediator.cs
namespace Architecture.Shell.Cqrs;

/// <summary>
/// Single entry point for sending commands and queries.
/// Routes requests to handlers via configured pipeline behaviors.
/// </summary>
/// <remarks>
/// <para><strong>Behavioral Guarantees:</strong></para>
/// <list type="bullet">
/// <item>Each request type maps to exactly ONE handler (validated at startup)</item>
/// <item>Pipeline behaviors execute in configured order</item>
/// <item>Cancellation tokens propagate through entire pipeline</item>
/// <item>Exceptions propagate to caller for infrastructure error handling</item>
/// <item>Result&lt;T&gt; returns for business error handling</item>
/// </list>
/// </remarks>
public interface IMediator
{
    /// <summary>
    /// Sends a request that returns a response.
    /// </summary>
    /// <typeparam name="TResponse">The response type.</typeparam>
    /// <param name="request">The request to send (command or query).</param>
    /// <param name="cancellationToken">Cancellation token for early termination.</param>
    /// <returns>The response after processing through the pipeline.</returns>
    /// <exception cref="InvalidOperationException">When zero or multiple handlers registered for request type.</exception>
    /// <exception cref="OperationCanceledException">When cancellation is requested.</exception>
    Task<TResponse> SendAsync<TResponse>(
        IBaseRequest request,
        CancellationToken cancellationToken);
}

File: typescript-nodejs/architecture-shell-cqrs/src/IMediator.ts
import { BaseRequest } from './BaseRequest';

/**
 * IMediator - Single entry point for sending commands and queries
 *
 * Routes requests to handlers via configured pipeline behaviors.
 * Provides centralized dispatcher for all application-layer operations.
 *
 * @remarks
 * Behavior:
 * 1. Resolve handler for request type (fail if zero or multiple handlers registered)
 * 2. Resolve applicable pipeline behaviors (filtered by type guards)
 * 3. Build behavior chain in configured order
 * 4. Execute pipeline → behaviors wrap handler execution
 * 5. Return result or propagate exception
 *
 * Constraints:
 * - Handler resolution MUST occur at DI container build time (not runtime reflection)
 * - Pipeline construction MUST happen per-request (behaviors may be stateful per request)
 *
 * @example
 * ```typescript
 * // Sending a command
 * const command = new CreateOrderCommand(customerId, items);
 * const result = await mediator.send(command, signal);
 * if (result.isSuccess) {
 *   console.log(`Order created: ${result.value}`);
 * }
 *
 * // Sending a query
 * const query = new GetOrderDetailsQuery(orderId);
 * const dto = await mediator.send(query, signal);
 * ```
 *
 * @see {@link IPipelineBehavior} - For implementing cross-cutting concerns
 * @see {@link IRequestHandler} - For implementing request handlers
 */
export interface IMediator {
  /**
   * Sends a request with return value through the mediator pipeline
   *
   * @typeParam TResponse - The response type
   * @param request - The request to send (must extend BaseRequest<TResponse>)
   * @param signal - Cancellation signal for early termination
   * @returns Promise resolving to the response
   * @throws If no handler registered or handler/behavior throws exception
   */
  send<TResponse>(
    request: BaseRequest<TResponse>,
    signal: AbortSignal
  ): Promise<TResponse>;

  /**
   * Sends a request without return value (void) through the mediator pipeline
   *
   * @param request - The request to send (must extend BaseRequest)
   * @param signal - Cancellation signal for early termination
   * @returns Promise resolving when request is processed
   * @throws If no handler registered or handler/behavior throws exception
   */
  send(request: BaseRequest, signal: AbortSignal): Promise<void>;
}
